docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
+  }
+}
+
+// Already exists in browser-storage-sync.
+interface PlacesHistoryStorage {
+  fun getTopFrecentSites(num: Int)
+}
+
+class DefaultTopSitesStorage(
+  val pinnedSitesStorage: PinnedSitesStorage,
+  val historyStorage: HistoryStorage? = null
+) : TopSitesStorage {
+
+  /**
+   * Merge data sources here, return a single list of top sites.
+   */
+  override fun getTopSites(numFrecent): List<TopSite>
+}
+
+/**
+ * Use cases can be used for adding a top from different places like a context menu.
+ */
+class TopSitesUseCases(topSitesStorage: TopSitesStorage) {
+  val addTopSites: AddTopSiteUseCase
+  val removeTopSites: RemoveTopSiteUseCase
+}
+/**
+ * View-bound feature that updates the UI when changes are made.
+ */
+class TopSitesFeature(
+  val storage: TopSitesStorage,
+  val presenter: TopSitesPresenter,
+  val view: TopSitesView,
+  val defaultSites: () -> List<TopSites>
+) {
+  fun start()
+  fun stop()
+}
+```
+
+## Drawbacks
+
+* `TopSitesStorage` always pulls sites from a persistent storage (Room). This could cause an excessive amount of reads from disk - this is drawback that exists in today's implementation as well. Places has an in-memory cache that should not be affected by this.
```

Comment:
It's fine to get this from the db, caching layers add complexity that is very nice to avoid if you don't need it.

Also, I don't think places has an in-memory cache, it reads data from the sqlite db. It's _ really quick_, since that's a `select` against a table with an index on the `frecent` score, IIRC.

Reply:
Unfortunately, I believe there are already some startup perf concerns with reading databases that need to show data on the homescreen, so we may run into this sooner than later 🤷 . For example: https://github.com/mozilla-mobile/android-components/issues/7459

Reply:
From that discussion, it feels like the `LiveData` part was problematic, not the db access itself?

But yeah, while you'll need to get this data from somewhere on first load, subsequent loads can benefit from an in-memory cache to fill out the UI sooner. This logic can be hidden within the feature later on (or live in the app); for example, we can compute the top sites now and then as a bg job, keep our in-memory cache updated, etc... I wouldn't worry about this for now, doesn't seem like anything here will get in the way of this sort of thing.

Reply:
We can add a read-through cache to the Places connection for top sites, but I think that first read on startup that populates the cache is always going to be a potential problem, no? Regardless of whether you're reading from the `Room` SQLite DB or Places SQLite DB?

Reply:
I'm leaning towards Grisha's thoughts - let's optimize for perf later until we know how we want this to function.

There are some [other discussions here](https://github.com/mozilla-mobile/fenix/pull/13009) about using different cache types to satisfy the visual completeness that UX is looking for. So yes, the initial read will always hurt us, the type of cache is what we can figure out later or even worse, how many layers of it (this last minute is a joke, please no :).